File: backend/app/routers/appendices.py
"""Appendix-document REST endpoints (Phase 5.4).

CRUD operations on the per-HOA appendix manifest. Thin router that
delegates business logic to ``appendix_service``; auth applied per
endpoint via ``get_current_user``.
"""

from __future__ import annotations

from typing import Literal, Optional

from fastapi import APIRouter, Depends, File, Form, HTTPException, UploadFile
from fastapi.responses import FileResponse
from pydantic import BaseModel
from sqlalchemy.orm import Session

from ..ai_implementation.db import get_session
from ..auth.dependencies import get_current_user
from ..disclosure_package.appendix_storage import appendix_file_path
from ..services.appendix_service import (
    AppendixDocumentResponse,
    AppendixNotFound,
    PropertyNotFound,
    get_appendix,
    list_appendices,
    retire_appendix,
    update_appendix,
    upload_appendix,
)


router = APIRouter(tags=["Appendices"])


def _actor_email(actor: dict) -> str:
    return str(actor.get("email") or actor.get("name") or "unknown")


class AppendixUpdateRequest(BaseModel):
    """Body for PUT /hoa/{hoa_id}/appendices/{appendix_id}.

    Every field is optional — the service applies just the supplied
    keys. ``expected_version`` is required for the optimistic-lock
    check; the frontend gets it from the GET response and sends it
    back on every PUT.
    """

    expected_version: int
    display_title: Optional[str] = None
    default_display_order: Optional[int] = None
    include_by_default: Optional[bool] = None
    cadence: Optional[Literal["persistent", "annual", "one_time"]] = None
    annual_year: Optional[int] = None
    valid_through_year: Optional[int] = None
    required_flag: Optional[bool] = None
    needs_cadence_review: Optional[bool] = None
    # null | "insurance" — omit field to leave unchanged; send null to clear
    package_role: Optional[str] = None
    package_role_set: Optional[bool] = None


@router.get(
    "/hoa/{hoa_id}/appendices", response_model=list[AppendixDocumentResponse]
)
def list_hoa_appendices(
    hoa_id: int,
    include_retired: bool = False,
    session: Session = Depends(get_session),
    current_user: dict = Depends(get_current_user),
) -> list[AppendixDocumentResponse]:
    raw_conn = session.connection().connection
    return list_appendices(
        property_id=hoa_id,
        connection=raw_conn,
        include_retired=include_retired,
    )


@router.post(
    "/hoa/{hoa_id}/appendices", response_model=AppendixDocumentResponse
)
async def upload_hoa_appendix(
    hoa_id: int,
    file: UploadFile = File(...),
    display_title: str = Form(...),
    cadence: Literal["persistent", "annual", "one_time"] = Form("persistent"),
    annual_year: Optional[int] = Form(None),
    valid_through_year: Optional[int] = Form(None),
    required_flag: bool = Form(False),
    include_by_default: bool = Form(True),
    package_role: Optional[str] = Form(None),
    session: Session = Depends(get_session),
    current_user: dict = Depends(get_current_user),
) -> AppendixDocumentResponse:
    file_bytes = await file.read()
    raw_conn = session.connection().connection
    try:
        return upload_appendix(
            property_id=hoa_id,
            file_bytes=file_bytes,
            original_filename=file.filename or "appendix.pdf",
            display_title=display_title,
            cadence=cadence,
            annual_year=annual_year,
            valid_through_year=valid_through_year,
            required_flag=required_flag,
            include_by_default=include_by_default,
            package_role=package_role,
            uploaded_by=_actor_email(current_user),
            connection=raw_conn,
        )
    except PropertyNotFound as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc


@router.get("/hoa/{hoa_id}/appendices/{appendix_id}/download")
def download_hoa_appendix(
    hoa_id: int,
    appendix_id: int,
    session: Session = Depends(get_session),
    current_user: dict = Depends(get_current_user),  # noqa: ARG001
) -> FileResponse:
    """Download the file for an appendix document."""
    raw_conn = session.connection().connection
    try:
        doc = get_appendix(property_id=hoa_id, appendix_id=appendix_id, connection=raw_conn)
    except AppendixNotFound as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    path = appendix_file_path(doc.file_id)
    if not path.exists():
        raise HTTPException(status_code=404, detail="Appendix file not found on disk")
    return FileResponse(path=path, filename=doc.file_name, media_type="application/pdf")


@router.put(
    "/hoa/{hoa_id}/appendices/{appendix_id}",
    response_model=AppendixDocumentResponse,
)
def update_hoa_appendix(
    hoa_id: int,
    appendix_id: int,
    payload: AppendixUpdateRequest,
    session: Session = Depends(get_session),
    current_user: dict = Depends(get_current_user),
) -> AppendixDocumentResponse:
    raw_conn = session.connection().connection
    try:
        # Pydantic: package_role in body means set (including null to clear)
        role_set = (
            payload.package_role_set
            if payload.package_role_set is not None
            else "package_role" in payload.model_fields_set
        )
        return update_appendix(
            property_id=hoa_id,
            appendix_id=appendix_id,
            expected_version=payload.expected_version,
            display_title=payload.display_title,
            default_display_order=payload.default_display_order,
            include_by_default=payload.include_by_default,
            cadence=payload.cadence,
            annual_year=payload.annual_year,
            valid_through_year=payload.valid_through_year,
            required_flag=payload.required_flag,
            needs_cadence_review=payload.needs_cadence_review,
            package_role=payload.package_role,
            package_role_set=bool(role_set),
            connection=raw_conn,
        )
    except AppendixNotFound as exc:
        # 409 on version mismatch, 404 otherwise
        if "version mismatch" in str(exc):
            raise HTTPException(status_code=409, detail=str(exc)) from exc
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc


@router.delete(
    "/hoa/{hoa_id}/appendices/{appendix_id}",
    response_model=AppendixDocumentResponse,
)
def retire_hoa_appendix(
    hoa_id: int,
    appendix_id: int,
    session: Session = Depends(get_session),
    current_user: dict = Depends(get_current_user),
) -> AppendixDocumentResponse:
    """Retire (soft-delete) the appendix — prior packages still
    reference the row via ``annual_package_appendices``. A hard delete
    would orphan those, so we mark ``status='retired'`` instead.
    """
    raw_conn = session.connection().connection
    try:
        return retire_appendix(
            property_id=hoa_id,
            appendix_id=appendix_id,
            connection=raw_conn,
        )
    except AppendixNotFound as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc

File: frontend/src/app/api/appendices.ts
// Typed API client for the per-HOA appendix manifest (Phase 5.4 of
// dre-driven-assessment-engine).
//
// Endpoints (backend at backend/app/routers/appendices.py):
//   GET    /hoa/{hoa_id}/appendices                          → AppendixDocument[]
//   POST   /hoa/{hoa_id}/appendices  (multipart)             → AppendixDocument
//   PUT    /hoa/{hoa_id}/appendices/{appendix_id}            → AppendixDocument
//   DELETE /hoa/{hoa_id}/appendices/{appendix_id}            → AppendixDocument (retired)

import { BASE_URL } from './config';
import { authHeaders, handleResponse } from './http';

export type AppendixCadence = 'persistent' | 'annual' | 'one_time';
export type AppendixStatus = 'active' | 'retired';

export interface AppendixDocument {
  appendix_id: number;
  property_id: number;
  file_id: string;
  file_name: string;
  display_title: string;
  default_display_order: number;
  required_flag: boolean;
  include_by_default: boolean;
  cadence: AppendixCadence;
  annual_year: number | null;
  valid_through_year: number | null;
  needs_cadence_review: boolean;
  status: AppendixStatus;
  version_int: number;
  // H7: backing file is gone from storage — re-upload or deselect before it
  // blocks a render.
  file_missing?: boolean;
}

export interface AppendixUpdateRequest {
  expected_version: number;
  display_title?: string;
  default_display_order?: number;
  include_by_default?: boolean;
  cadence?: AppendixCadence;
  annual_year?: number | null;
  valid_through_year?: number | null;
  required_flag?: boolean;
  needs_cadence_review?: boolean;
}

export async function listAppendices(
  hoaId: number,
  options: { includeRetired?: boolean } = {},
): Promise<AppendixDocument[]> {
  const params = new URLSearchParams();
  if (options.includeRetired) params.set('include_retired', 'true');
  const qs = params.toString();
  const url = `${BASE_URL}/hoa/${hoaId}/appendices${qs ? `?${qs}` : ''}`;
  const res = await fetch(url, { headers: authHeaders() });
  return handleResponse<AppendixDocument[]>(res);
}

export async function uploadAppendix(
  hoaId: number,
  args: {
    file: File;
    displayTitle: string;
    cadence?: AppendixCadence;
    annualYear?: number | null;
    validThroughYear?: number | null;
    requiredFlag?: boolean;
    includeByDefault?: boolean;
  },
): Promise<AppendixDocument> {
  const form = new FormData();
  form.append('file', args.file);
  form.append('display_title', args.displayTitle);
  form.append('cadence', args.cadence ?? 'persistent');
  if (args.annualYear != null) form.append('annual_year', String(args.annualYear));
  if (args.validThroughYear != null)
    form.append('valid_through_year', String(args.validThroughYear));
  if (args.requiredFlag != null) form.append('required_flag', String(args.requiredFlag));
  if (args.includeByDefault != null)
    form.append('include_by_default', String(args.includeByDefault));

  const res = await fetch(`${BASE_URL}/hoa/${hoaId}/appendices`, {
    method: 'POST',
    headers: authHeaders(),
    body: form,
  });
  return handleResponse<AppendixDocument>(res);
}

export async function updateAppendix(
  hoaId: number,
  appendixId: number,
  body: AppendixUpdateRequest,
): Promise<AppendixDocument> {
  const res = await fetch(`${BASE_URL}/hoa/${hoaId}/appendices/${appendixId}`, {
    method: 'PUT',
    headers: { ...authHeaders(), 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
  return handleResponse<AppendixDocument>(res);
}

export async function retireAppendix(
  hoaId: number,
  appendixId: number,
): Promise<AppendixDocument> {
  const res = await fetch(`${BASE_URL}/hoa/${hoaId}/appendices/${appendixId}`, {
    method: 'DELETE',
    headers: authHeaders(),
  });
  return handleResponse<AppendixDocument>(res);
}

export async function downloadAnnualAppendix(
  hoaId: number,
  appendixId: number,
  fileName: string,
): Promise<void> {
  const res = await fetch(`${BASE_URL}/hoa/${hoaId}/appendices/${appendixId}/download`, {
    headers: authHeaders(),
  });
  if (!res.ok) throw new Error(`Download failed: ${res.status}`);
  const blob = await res.blob();
  const url = URL.createObjectURL(blob);
  const a = document.createElement('a');
  a.href = url;
  a.download = fileName;
  a.click();
  URL.revokeObjectURL(url);
}
